### PR TITLE
[query] Add interfaces for lowering distributed sorts

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 1]
```diff
@@ -186,6 +186,34 @@ object LowerTableIR {
...
           loweredChild.copy(body = StreamMap(loweredChild.body, row.name, Subst(newRow, BindingEnv(env, scan = Some(env)))))
 
+        case t@TableKeyBy(child, newKey, isSorted: Boolean) =>
+          val loweredChild = lower(child)
+          val nPreservedFields = loweredChild.partitioner.kType.fieldNames
+            .zip(newKey)
+            .takeWhile { case (l, r) => l == r }
+            .length
+
+          require(!isSorted || nPreservedFields > 0 || newKey.isEmpty)
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:196]: I don't know why this is necessary.  If n = 0, you should just have empty key intervals.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 2]
```diff
@@ -186,6 +186,34 @@ object LowerTableIR {
...
+
+          if (nPreservedFields == newKey.length)
+            loweredChild
+          else if (isSorted) {
+            val newPartitioner = loweredChild.partitioner
+              .coarsen(nPreservedFields)
+              .extendKey(t.typ.keyType)
+            if (newPartitioner.isStrict)
+              loweredChild.copy(partitioner = newPartitioner)
+            else throw new LowererUnsupportedOperation("TableKeyBy with isSorted but a non-strict partitioner requires partitioner adjustments")
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:206]: I'm confused why this is necessary.  I thought partitioners allowed the endpoints to overlap (and some downstream consumer would have to sort it out).

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 3]
```diff
@@ -186,6 +186,34 @@ object LowerTableIR {
...
+          } else
+            ctx.backend.lowerDistributedSort(loweredChild, newKey.map(k => SortField(k, Ascending)))
+
+        case TableOrderBy(child, sortFields) =>
+          val loweredChild = lower(child)
+          if (TableOrderBy.isAlreadyOrdered(sortFields, loweredChild.partitioner.kType.fieldNames))
+            loweredChild.copy(partitioner = loweredChild.partitioner.coarsen(0))
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:213]: I don't think you need to coarsen here.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:196]: this requirement was lifted from RVD. I think it's correct -- if n == 0 (and isSorted == true) then you'll end up with a keyed dataset where you don't have any partitioning information. You can't do anything with that.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:206]: I don't think so. We could redesign things in TableStage.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:213]: I guess I need to clarify what the relationship of a partitioner is to the key of a lowered TIR. Does the partitioner key type have to be the same? Longer?

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:213]: I though the partitioner could be longer (you potentially let it stay longer in the `nPreservedFields == newKey.length` case above), but it seems not.  From RVD:

 > require(typ.kType.virtualType isIsomorphicTo partitioner.kType)

(It seems isIsomorphicTo is equals for base structs: it allows structs and tuples to compare equal.)

So why isn't this `coarsen(sortFields.length)`?

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala (hunk 1 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:196]: I can't find any evidence that overlap isn't allowed in RVD.  

I also don't buy n = 0 is special.  If you're coarsening the partitioner, distinct partition endpoints can become the same.  For example: [a 0, a 5], [a 6, a 11] with coarsen(1).  The only thing I could find was this:

 > lteqWithOverlap(allowedOverlap)(left.right, right.left)

which appears to allow equality up to allowed overlap (which appears to be the key length - 1 in the initial case).

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:213]: TableOrderBy can sort in descending order. We'll either need to encode ascending/descending into the partitioner, or erase sortedness when we order by (descending).

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:213]: er, nevermind, I think you're right -- no need to coarsen.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 4]
```diff
@@ -55,6 +55,12 @@ abstract class TableStage(
       broadcastVals.foldLeft(partition(ctx))((accum, name) => Let(name, GetField(glob, name), accum)))
     if (bind) wrapInBindings(cda) else cda
   }
+
+  def castPartitioner(newPartitioner: RVDPartitioner): TableStage = {
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:59]: This is not a cast.  It is just a new partitioner.  `with[New]Partitioner`?  `setPartitioner`?  `copy`?

**Comment by tpoterba** [comment]
Let's dig into the n=0 case first.

I think pretty strongly that if nPreservedFields==0, you've done something very wrong. 

This means you've done a key cast that leaves zero information, so how could you possibly join this with another table? You'd have to create a single partition, and we'd never want to do that.

**Comment by cseed** [comment]
> I think pretty strongly that if nPreservedFields==0, you've done something very wrong.

I don't disagree, but if you're concerned about this, I think an assert here is both in the wrong place and the wrong solution.

I think for the design the IR, we should focus on (1) simple operations (this is already more complicated than I'd like), (2) that are composable, and (3) minimize special cases.

By composability, I mean each IR should have a (local) contract, and each program composed from that local contract should be valid.  Breaking this introduces a lot of potential bugs that can't be reasoned about locally, which is not good.   The IR nodes do what they do.  If you don't like what they do, we should probably find different ones.

 > so how could you possibly join this with another table? You'd have to create a single partition, and we'd never want to do that.

Yeah, create a single partition.  Or reshuffle if the partitioner has too little information.  How much is too little?  What if nPreservedFields==1 and we're down to 1 partition?  Should that be an error?  2 partitions?  How many partitions is too few?  Any time nPreservedFields is less than the requested keys, you could get down to 1 partition.  This is a continuous issue and rejecting the extreme case doesn't actually solve the problem.

I guess isSorted isn't user exposed, but this seems dangerously close to reporting a user error with an assertion.  When the service comes up, hopefully not too long, we're going to want to document the IR and make it public.  So if we want to reject this case, we should do it early on: when the IR is parsed and/or constructed.  (In general I think to give a nice experience we're going to have to do more up-front validation.)  This is what I mean when I say "find another IR".

In summary:
 - If we're going to have this assertion, it needs to be in TableKeyBy constructor, and
 - This is a complex and serious issue that isn't actually solved by your assertion.

**Comment by tpoterba** [comment]
OK, thanks for the pushback. This is ready for another look. Here is my understanding of the semantics of partitioners and preserved keys:

1. A TableStage from a lowered TableIR must have a partitioner that begins with the table key, but the partitioner can have a longer key. That partitioner need not be strict; consumers that require strictness like TableAggregateByKey or TableDistinct will strictify their lowered children.

2. The lowering rule for TableKeyBy makes no assertion about the number of preserved key fields. It is possible for client code to pass a TableKeyBy IR with  0 preserved key fields, which will result in a correct but possibly unperformant execution. We will document this when we document the IR system in several months.

**Comment by cseed** [comment]
I agree, except:

 > which will result in a correct but possibly unperformant execution

Right now, yes, but this is because we make a bad decision (1 partition) when strictifying partitions with no information and what we should do is make a better decision.
